[pkg/stanza] [receiver/windowseventlogreceiver]: speed up receiver

MrAnno opened the pull request:
#### Description

With the default `max_reads: 100`, `poll_interval: 1s` config fields, the receiver could process a maximum of 100 events per second.

This PR improves the performance of the Windows Event Log receiver by introducing an interruptible `readAll()` method, which tries to read all the messages in  `max_reads`-sized batches and only falls back to polling when it reaches the end of the eventlog channel. The timer is started after each read cycle from now on.

Note that the same performance cannot be achieved by simply adjusting `max_reads`, because `RPC_S_INVALID_BOUND` errors limit the maximum configurable batch size. Alternatively, setting `poll_interval` to a smaller value would also incur some overhead.

MrAnno committed:
[pkg/stanza] [receiver/windowseventlogreceiver]: speed up receiver

With the default `max_reads: 100`, `poll_interval: 1s` config fields,
the receiver could process a maximum of 100 events per second.

This commit introduces an interruptible readAll() method, which tries to
read all the messages in max_reads-sized batches and only falls back to
polling when it reaches the end of the eventlog channel.

Note that the same performance cannot be achieved by simply adjusting
max_reads, because RPC_S_INVALID_BOUND errors limit the maximum
configurable batch size. Alternatively, setting poll_interval to a smaller
value would also incur some overhead.

pjanotti reviewed (commented):
Thanks for your contribution @MrAnno!

I would describe this change as removing the upper-bound per poll and collecting events until there are no more events to be consumed. This can be useful in various scenarios. That said I'm a bit worried that now there are no upper bounds to the receiver reading events. It seems a good defense to have some default upper bound - it can be time or max number of events, I'm not sure at this point. Such upper bound limit can avoid the cases where something is generating a flood of events (and probably already using lots of CPU) and the collector is not backing off at all, causing CPU oversubscription on the box.

Consider the case that the collector is also configured to collect all events, it should back off from time to time even if that is going to make it taking longer to read all events. That said this change will make it catchup much faster in such cases.

MrAnno committed:
[pkg/stanza] [receiver/windowseventlogreceiver]: introduce rate-limit

MrAnno commented:
@pjanotti Thank you for the quick response.

I've added a new config option called `rate_limit` so that the user has control over the mentioned upper bound expressed in `events/sec`.

I think setting the default value to anything other than 0 would be pretty arbitrary, because everything we want to achieve here depends on the given system's resources (for a PC, 1000 is a good default, but on a real collector server, I probably wouldn't go under 20 000).

If my speed-up patch is considered a breaking change due to its possible peak resource consumption, we could come up with a fair default. Otherwise, I would prefer 0, because the CPU usage of `otelcontribcol` stayed under 40% even when I started flooding events with a custom C tool (50.000 events/sec sent into the event viewer on 4 ordinary CPU cores).

pjanotti reviewed (commented):
@MrAnno I'm fine with defaulting to no rate limit by default, but, I think that just counting the overall number of events read in a single call to `readAll` and letting `readThrottle` return false when that limit is reached achieves the effect that we want. Simple having the limit per poll is easier for users to understand and control.

MrAnno committed:
[pkg/stanza] [receiver/windowseventlogreceiver]: replace rate_limit with max_events_per_poll

pjanotti reviewed (commented):
Sorry for the delay @MrAnno - I think it is almost there, some small things and the need for a test, at least when the limiting is not zero.

pjanotti commented:
@MrAnno noticed that I merged main into this PR branch, so remember to pull before you make any changes. My understanding is that you are looking into adding a basic test for the new feature, it will also be good to have some config tests.

pjanotti reviewed (commented):
Remaining issues:

1. https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/18648688527/job/53211077377?pr=43195#step:6:168
2. https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/18648688527/job/53211077377?pr=43195#step:8:12

Could you please update the description of the PR so it matches the current (hopefully final) implementation?

MrAnno commented:
Sorry, I forgot about this. I'll try to fix the test soon.

MrAnno committed:
[pkg/stanza]: allow monkey patching typed evt* functions

MrAnno committed:
[pkg/stanza]: fix tests

It was impossible to implement proper mocks for evt* functions by
patching the LazyProc calls directly, as runtime/checkptr.go has too
strict checks for that.

Now, the evt* functions themselves are patched, which have the proper
type parameters.

MrAnno commented:
@pjanotti Sorry for the delay. I think the test is finally ready as well.

MrAnno commented:
Sorry again. I'm finally on vacation, I'm trying to fix the last remaining issue now.

MrAnno commented:
Hm, it seems that the goleak error about the unexpected goroutine was not introduced by my PR, I can see the same error with a slightly different stack trace on `main` in my environment:

```
PS C:\Users\Anno\Desktop\opentelemetry-collector-contrib\pkg\stanza\operator\input\windows> go test -gcflags=-d=checkptr -run ^TestInputStart_RemoteSessionWithDomain$
PASS
goleak: Errors on successful test run: found unexpected goroutines:
[Goroutine 35 in state select, with github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/input/windows.(*Input).readOnInterval on top of the stack:
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/input/windows.(*Input).readOnInterval(0xc0001da200, {0x7ff736f8af98, 0xc0001922d0})
        C:/Users/Anno/Desktop/opentelemetry-collector-contrib/pkg/stanza/operator/input/windows/input.go:209 +0x117
created by github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/input/windows.(*Input).Start in goroutine 34
        C:/Users/Anno/Desktop/opentelemetry-collector-contrib/pkg/stanza/operator/input/windows/input.go:169 +0x9e8
]
exit status 1
FAIL    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/input/windows     3.237s
```

@pjanotti  Can you take a quick look, please?

pjanotti committed:
Merge branch 'main' into eventlog-speedup

pjanotti reviewed (approved):
Thanks for your diligent work on this @MrAnno!
